### PR TITLE
Feat/add features and loss functions

### PR DESCRIPTION
**Pull request by almahova**
## Summary
Adds differentiable spatial and temporal feature (`pyFDN.features`), and integrates Auraloss time-domain loss functions(`pyFDN.train.losses`).

## Changes
- **Auraloss Loss Functions**: Implemented `AuralossResponseLoss` adapter to bridge pyFDN's API and Auraloss's. 
- **Time-Domain Losses**: Added `MatchESR`, `MatchSISDR`, `MatchDC`, `MatchLogCosh`, `MatchSNR`, and `MatchSDSDR`.
- **Feature Extractors**: Added `mimo_rir_eigenvalues_per_frequency`, `energy_decay_curve` (EDC)
- **Tests**: Added unit test for features and losses

## Testing
- [x] `pytest` passes locally
- [x] New tests added for new functionality
- [ ] Coverage not decreased (`pytest --cov=pyFDN`)

## Checklist
- [ ] Pre-commit hooks pass (`pre-commit run --all-files`)
- [x] Docstrings updated for any changed public API
- [x] `HISTORY.rst` updated if this is a user-facing change

### CODE REVIEW

**Comment by SebastianJiroSchlecht** [comment]
Suggestion: could we reduce the boilerplate across the `Match…` losses by composing feature extraction, comparison, and reduction?

For example:

```python
Match(
    target=reference_ir,
    feature=Magnitude(),
    distance=SquaredError(),
    reduction=Mean(),
)
```

`Match` would inherit from `ResponseLoss` and own reference alignment (length, channels, device, dtype) and feature caching. A response feature would extract the waveform, magnitude, phase, EDC, mel representation, etc. The distance would compare the features, and the reduction would aggregate the result into a scalar.

I would distinguish distance from reduction: L1/L2 specify how values differ, whereas mean/sum/median specify how errors are aggregated. Existing names such as `MatchMagnitude` could remain thin convenience wrappers. Phase matching would use a circular distance rather than ordinary subtraction.

Also, please avoid introducing Auraloss as a dependency. For this change, the requested reimplementation scope is only the STFT loss, `MultiResolutionSTFTLoss`, and the mel-scaled version, implemented directly using the existing PyTorch stack—not the six Auraloss time-domain loss wrappers. These spectral losses can share the feature/comparison machinery, with explicit handling of their spectral-convergence and log-magnitude terms and aggregation across resolutions.

Please implement from the mathematical definitions, document the conventions, and test numerical behavior and gradients, including silent inputs and MIMO shapes. Auraloss can be used for one-off numerical comparison during development without becoming a runtime dependency.

**Comment by almahova** [comment]
@lsbharadwaj - I have made a few changes, including creating new Match class in order to avoid some repeated code in the loss-functions. I've also deleted the Auraloss library use.
Can I ask your kind review? do you have any more ideas for improvement and centralize 'Match..'? 
Would you agree with that implementation?

**Comment by lsbharadwaj** [comment]
@almahova I continued on the same lines as you. I converted some more functions as per your implementations. There are still a few functions that need to be converted. May be you can do a few and I will paralelly do others. Also we can later consider moving Reductions, distances and features into separate files. Currently all are in match.py

**Comment by almahova** [comment]
@lsbharadwaj 
The classes that currently don't inharit from Match are: 

spectral.py: 
1. AsymmetricFlatMagnitude(ResponseLoss)
2. _FlamoSpectrogramLoss(ResponseLoss)
3. MatchSpectrogram(_FlamoSpectrogramLoss)
4. MatchMelSpectrogram(_FlamoSpectrogramLoss)

temporal.py:
1. Energy(ResponseLoss)
2. MatchEnergyDecay(ResponseLoss)
3. MatchCumulativeEnergy(ResponseLoss)

The Match pipline takes 2 signals, h (the output from the model) and h_target (our reference), in the same type of representation, and compare them, leading to a scalar loss. 

The losses that mentioned above don't fit into that pattern, for example: not using h_target, or having multiple passes.
As far as I concered, it's better to keep the inharitance from ResponseLoss. 

@SebastianJiroSchlecht - Do you agree with keeping these as standalone ResponseLoss implementations?

**Comment by SebastianJiroSchlecht** [comment]
Review of ef6e6d10 (fix channel indexing in `SpectrogramFeature` / `PhaseSpectrogramFeature`)

The reshape fix is correct: `signal` is flattened out-major via `permute(1, 2, 0)`, so `stacked` is already `(n_nfft, out·in, freq[, frames])` and a plain reshape restores `(out, in)`. The old `permute` scrambled channels in `SpectrogramFeature` and entangled frequency with frames in `PhaseSpectrogramFeature`, even for a single channel.

The tests are good: I ran them against the previous commit and 3 of 11 fail (per-channel identity, direct `torch.angle(torch.stft(...))`, and the quarter-cycle check), all 11 pass with the fix. So they are real regression tests.

Two follow-ups, not blocking this commit:
1. `FlatMagnitude` and `SpectralFlatness` pass `dummy_target = np.array([1.0])` into `Match`, while `FlatSpectrogram` already passes `target=None`. Please pick one: let `Match` accept `target=None` and remove the dummy arrays.
2. The scope from my earlier comment is still open: `MatchSpectrogram` and `MatchMelSpectrogram` still wrap FLAMO's `mss_loss` / `mel_mss_loss`. The native multi-resolution STFT loss (spectral convergence + log-magnitude) and its mel variant, with tests for gradients, silent inputs and MIMO shapes, are still to do.

On the standalone question: keeping `Energy`, `AsymmetricFlatMagnitude`, `MatchEnergyDecay` and `MatchCumulativeEnergy` as plain `ResponseLoss` is fine. The two spectrogram losses are the exception, see point 2.
